File: csharp/UrakawaToolkit/undo/CommandFactory.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace urakawa.undo
{
	/// <summary>
	/// Factory for creating <see cref="ICommand"/>s
	/// </summary>
	public class CommandFactory : WithPresentation
	{
		/// <summary>
		/// Creates a <see cref="ICommand"/> matching a given Xuk QName
		/// </summary>
		/// <param name="xukLocalName">The local name part of the Xuk QName</param>
		/// <param name="xukNamespaceUri">The namespace uri part of the Xuk QName</param>
		/// <returns>The created command or <c>null</c> if the Xuk QName is not recognized</returns>
		public ICommand createCommand(string xukLocalName, string xukNamespaceUri)
		{
			if (xukNamespaceUri == ToolkitSettings.XUK_NS)
			{
				if (xukLocalName == typeof(CompositeCommand).Name)
				{
					return createCompositeCommand();
				}
			}
			return null;
		}

		/// <summary>
		/// Creates a <see cref="CompositeCommand"/>
		/// </summary>
		/// <returns>The created composite command</returns>
		public CompositeCommand createCompositeCommand()
		{
			CompositeCommand newCmd = new CompositeCommand();
			newCmd.setPresentation(getPresentation());
			return newCmd;
		}
	}
}

File: trunk/java/UMLGraph/src/org/daisy/urakawa/undo/CommandFactory.java
package org.daisy.urakawa.undo;

import org.daisy.urakawa.WithPresentation;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.xuk.XukAble;

/**
 * <p>
 * This is the factory that creates {@link org.daisy.urakawa.undo.Command}
 * instances.
 * </p>
 * 
 * @depend - Create - org.daisy.urakawa.undo.Command
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 * @stereotype OptionalLeafInterface
 * @depend - Aggregation 1 org.daisy.urakawa.Presentation
 */
public interface CommandFactory extends XukAble, WithPresentation {
	/**
	 * <p>
	 * Creates a new Command
	 * </p>
	 * <p>
	 * This factory method takes arguments to specify the exact type of object
	 * to create, given by the unique QName (XML Qualified Name) used in the XUK
	 * serialization format. This method can be used to generate instances of
	 * subclasses of the base object type.
	 * </p>
	 * 
	 * @param xukLocalName
	 *            cannot be null, cannot be empty string.
	 * @param xukNamespaceURI
	 *            cannot be null, but can be empty string.
	 * @return can return null (in case the QName specification does not match
	 *         any supported type).
	 * @tagvalue Exceptions "MethodParameterIsNull-MethodParameterIsEmptyString"
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 * @throws MethodParameterIsEmptyStringException
	 *             Empty string '' method parameter is forbidden:
	 *             <b>xukLocalName</b>
	 */
	public Command createCommand(String xukLocalName, String xukNamespaceURI)
			throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException;

	/**
	 * @return an empty CompositeCommand object.
	 */
	public CompositeCommand createCompositeCommand();
}
